### PR TITLE
Replace SymbolProcessor#init with SymbolProcessorProvider interface

### PR DESCRIPTION
**Pull request by edrd-f**
This PR introduces a new way of instantiating `SymbolProcessor`s by moving `init` into a separate interface. This has the following benefits:

* Plugins are no longed forced to use `var`s. The resulting code is shorter and less error-prone:

```kotlin
class Processor(private val codeGenerator: CodeGenerator) : SymbolProcessor {
    override fun process(...) { ... }    

    class Provider : SymbolProcessorProvider {
        override fun create(codeGenerator: CodeGenerator, ...) = Processor(codeGenerator)
    }
}
```

* KSP is able to instantiate multiple `SymbolProcessor`s, opening the path for parallel execution of processors (e.g. by module).
* Temporal coupling between initialization of processors and lifecycle functions is eliminated.
* The other model (with `lateinit var`) can still be used by implementing both interfaces on the same class. For example:

```kotlin
class Processor : SymbolProcessor, SymbolProcessorProvider {
    lateinit var codeGenerator: CodeGenerator

    override fun create(codeGenerator: CodeGenerator, ...) = this.apply {
        this.codeGenerator = codeGenerator
    }

    override fun process(...) { ... }
}
```

The implementation allows plugins using different models (init and provider - see `InitPlusProviderIT`) on the same run, but if a projects registers both `SymbolProcessorProvider` and `SymbolProcessor` for service loading, initialization will occur twice.

### CODE REVIEW

File: gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/processor/TestSymbolProcessor.kt [hunk 1]
```diff
@@ -19,8 +19,9 @@ package com.google.devtools.ksp.gradle.processor
 import com.google.devtools.ksp.processing.CodeGenerator
 import com.google.devtools.ksp.processing.KSPLogger
 import com.google.devtools.ksp.processing.SymbolProcessor
+import com.google.devtools.ksp.processing.SymbolProcessorProvider
 
-abstract class TestSymbolProcessor : SymbolProcessor {
+abstract class TestSymbolProcessor : SymbolProcessor, SymbolProcessorProvider {
```
Comment (neetopia) [inline gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/processor/TestSymbolProcessor.kt]: Can you convert this processor into the new model you proposed where `var`s for stuff like `logger` and `codeGenerator` is converted to `val`s? Same for the `init-plus-provider` integration test.

**Review by neetopia — COMMENTED** [review]
Overall it looks good to me :) Thanks for the work. Can you do a rebase on master and send a PR to KSP repo? Would like keep it open on KSP repo for a while for some community feedback and merge it into master.

**Comment by edrd-f** [comment]
Hey @neetopia, thanks for taking the time to look at this. I'll update the branch and make the changes you mentioned as soon as possible.
Cheers.

File: gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/processor/TestSymbolProcessor.kt (hunk 1 above)
Comment (edrd-f) [reply gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/processor/TestSymbolProcessor.kt]: @neetopia, I updated the branch and converted `TestSymbolProcessor` to the new model in [this commit](https://github.com/edrd-f/ksp/pull/1/commits/9fd2db07314437a2acce9d63080f71b14f0476af). 

About the `init-plus-provider` integration test, I'm not sure I understand what you mean. The test uses both models, so the one with `init` uses `lateinit`, naturally.

**Comment by edrd-f** [comment]
Just opened it on google/ksp! :)

**Pull request closed without merging**